libraries/AP_Math/SCurve.cpp

```diff
@@ -61,8 +61,9 @@ void SCurve::calculate_track(const Vector3f &origin, const Vector3f &destination
 {
     init();
 
-    // leave track as zero length if origin and destination are the same
-    if (origin == destination) {
+    // leave track as zero length if the vector from origin to destination length squared is zero
+    const Vector3f direction = destination - origin;
+    if (is_zero(direction.length_squared())){
```

Comment:
If this is a problem then shouldn't it be fixed here:
template <typename T>
bool Vector3<T>::operator ==(const Vector3<T> &v) const
{
    return (is_equal(x,v.x) && is_equal(y,v.y) && is_equal(z,v.z));
}

Reply:
@peterbarker ?

Reply:
The '==' operator itself is correct. But the only way origin and destination get used eventually is conditioned upon 
` ( (origin - destination).length() )^2 < FLOAT_EPSILON` instead of what `==` checks which is 

`return ( ((origin.x -destination.x) < FLT_EPSILON) && ((origin.y -destination.y) < FLT_EPSILON) && .... );`

The issue is that the above `==` will pass because origin & destination differ by more than FLOAT_EPSILON. When I managed to hit it once while debugging the `(origin.z - desitnation.z) = -2.54035E-4` which is `>FLT_EPSILON` so it passes. 

But later in, the only place where origin and destination get used (now as direction), here
https://github.com/ArduPilot/ardupilot/blob/8b83b49e0b56751d3e96ea763680b43aaecb4cd4/libraries/AP_Math/control.cpp#L532-L534

kinematic_limit returns 0 as (-2.54035E-4)^2 < FLT_EPSILON

`vel_ max` and `accel_max` are then set to zero here,
https://github.com/ArduPilot/ardupilot/blob/8b83b49e0b56751d3e96ea763680b43aaecb4cd4/libraries/AP_Math/SCurve.cpp#L912-L916

Then because `vel_max = 0` & `accel_max = 0` we end at the Internal_Error here,
https://github.com/ArduPilot/ardupilot/blob/8b83b49e0b56751d3e96ea763680b43aaecb4cd4/libraries/AP_Math/SCurve.cpp#L79-L84

Hopefully that is a better explanation at what I think is going on. I'm not sure it would effect a vehicle unless data got stale.